amida-tech/amida-messaging-microservice
# Feature/threads privacy and filtering

If this PR fixes a bug, you _must_ add test cases representative of the bug.
#### What's this PR do?
1. Adds tests for the `threads` endpoints (not to be confused with the `thread` endpoints, which support a completely separate thread implementation)
2. Fixes a privacy/security issue that the `threads` endpoints had (any authenticated user could read any thread - even if they weren't included)
3. Adds a `start_date` filter to the "show thread" `threads` endpoint to allow fetching of only messages after the supplied `start_date`
4. Adds a `.eslintignore` file so that eslint will stop complaining about everything in `dist/`
#### Related JIRA tickets:
Orange-300
#### How should this be manually tested?
1. run `yarn test` and verify all tests pass
2. Via postman (or any other REST client of your choice!):
    * create three users (via auth service): user0, user1, and user2
    * as user0, create a thread between user0 and user1
      * `POST /threads { participants: ['user0', 'user1'], subject: '', topic: '', message: 'msg' }`
    * as user2, try to get messages of, view participants of, and reply to that thread
      * `GET /threads/thread/:threadId`
      * `GET /threads/thread/participants/:threadId`
      * `POST /threads/thread/:threadId/reply { message: 'reply' }`
      * all of these should return 404
3. Via postman (or any other REST client of your choice!):
    * Create a thread with a couple messages
      * `POST /threads { participants: ['user0', 'user1'], subject: '', topic: '', message: 'msg' }`
      * `POST /threads/thread/:threadId/reply { message: 'reply' }`
      * `POST /threads/thread/:threadId/reply { message: 'reply' }`
      * `POST /threads/thread/:threadId/reply { message: 'reply' }`
    * Get the messages of that thread using a start_date filter
      * `GET /threads/thread/:threadId?start_date=${date}`
      * verify only messages at or after the date you provided are returned
4. Run `yarn lint` and verify there are no linting errors
    * introduce a linting error in a random file
    * Run `yarn lint` and verify it finds the error you introduced
